draft-ietf-tls-esni.md

```diff
@@ -531,6 +531,57 @@ The decompression process in {{encoding-inner}} forbids
 "encrypted_client_hello" in OuterExtensions. This ensures the unauthenticated
 portion of ClientHelloOuter is not incorporated into ClientHelloInner.
 
+# Handshake Padding {#handshake-padding}
+
+Encrypting the ClientHelloInner is usually not enough to protect sensitive
+handshake parameters against traffic analysis. In particular, the length of the
+ciphertext, or the length of any encrypted handshake message that follows it,
+may also leak information about these parameters. To mitigate this issue for
+ClientHelloInner, the client implements the padding scheme specified in [[TODO:
+Point to the new padding mechanism introduced by PR#443]]. This leakage is
+addressed in the remainder of the handshake using the padding mechanism
+described here.
+
+Handshake padding is applied by both the client and backend server in case of
+ECH acceptance. The client determines if ECH was accepted using the procedure
+specified in {{handle-server-response}}. The backend server determines if ECH
+was accepted using the procedure specified in {{backend-server}}.
+
+A new handshake message is defined, which, in case of ECH acceptance, is sent by
+both the client and backend server immediately before the Finished message.
+
+~~~~
+    enum {
+        padding(16),
+    } HandshakeType;
+
+    struct {
+        uint8 zeros[length_of_padding];
+    } Padding;
+~~~~
+
+[[TODO: Address IANA considerations for new codepoint.]]
+
+`zeros` consists of any number of zero bytes. Upon receipt of a Padding message,
+a host MAY verify that the payload is correct (i.e., that it consists of all
+zeros) and abort with "illegal_parameter" if not. Each Padding message is
+incorporated into the handshake transcript just like any other handshake
+message.
+
+The Padding message MUST be sent by the client and backend server in case of ECH
+acceptance and it MUST immediately precede the Finished message. If a Padding
+message is received out-of-order, or it was received without ECH being accepted,
+then the receiver MUST treat it as an unexpected message.
+
+This mechanism is intended to allow the client and backend server to protect
+sensitive handshake parameters that would otherwise be leaked by the length of
+the encrypted handshake messages. Implementations must also take care to ensure
+that, when coalescing the handshake messages into a sequence of records, the
+lengths of the records are independent of the lengths of the messages.
```

Comment:
Sending padding as a separate message leaves open the possibility that some implementations could send the padding message in a separate record than the messages that need the padding to ensure indistinguishability. If that record can be reliably distinguished from the previous record (e.g. if an implementation always sends the padding and finished messages in the final handshake record from the server, and they are the only messages in that record), then the padding is no longer effective. 

I believe this scenario is addressed by the above warning about coalescing messages into a sequence of records, but I'm concerned it might be glossed over by implementors, especially if they send the messages one at a time without record-level padding assuming that the padding message is sufficient. Maybe instead it could be specified here that "implementors MUST coalesce the handshake messages up through and including the padding message into a sequence of records so that the length of the records are identical, with the exception of the last record which can have a shorter length"? While this is more restrictive, I think it may prevent some implementation errors down the road.